bindings/rust/s2n-tls/src/raw/config.rs

```diff
@@ -185,8 +274,79 @@ impl Builder {
         Ok(self)
     }
 
+    /// Set a custom callback function which is run after parsing the client hello.
+    ///
+    /// The callback can be called more than once.
+    pub fn set_client_hello_handler<T: 'static + ClientHelloHandler>(
+        &mut self,
+        handler: T,
+    ) -> Result<&mut Self, Error> {
+        unsafe extern "C" fn client_hello_cb(
+            connection_ptr: *mut s2n_connection,
+            context: *mut core::ffi::c_void,
+        ) -> libc::c_int {
+            let context = &mut *(context as *mut Context);
+            let handler = context.client_hello_handler.as_mut().unwrap();
+
+            let connection_ptr =
+                NonNull::new(connection_ptr).expect("connection should not be null");
+
+            // Since this is a callback, which receives a pointer to the connection, do
+            // not call drop on the connection object.
+            let mut connection = ManuallyDrop::new(Connection::from_raw(connection_ptr));
+
+            match handler.poll_client_hello(&mut connection) {
+                Poll::Ready(Ok(())) => {
+                    // FIXME: this is not always true and should be addressed in the final version
+                    // of the bindings. We set the extension since custom certificate loading is
+                    // the most common operation at the moment.
+                    //
+                    // A server that receives a client hello containing the "server_name" extension
+                    // MAY use the information contained in the extension to guide its selection of
+                    // an appropriate certificate to return to the client, and/or other aspects of
+                    // security policy.  In this event, the server SHALL include an extension of type
+                    // "server_name" in the (extended) server hello.
+                    s2n_connection_server_name_extension_used(connection_ptr.as_ptr())
+                        .into_result()
+                        .unwrap();
```

Comment:
Your FIXME implies you have at least one use case for not setting server_name_extension_used. Why set it initially? Won't that cause unnecessary pain when switching to the "final" version of the bindings? What's the issue with customers calling it from their callback?

Reply:
> Why set it initially?
"We set the extension since custom certificate loading is the most common operation at the moment." We expect most applications to set the certificate.

> What's the issue with customers calling it from their callback?
Having to call this function seems like a pretty sharp api (There is nothing forcing users to actually call this api and no way to enforce that its done correctly). So as part of the Rust bindings I wanted this to be the default.

> Won't that cause unnecessary pain when switching to the "final" version of the bindings?
I think you are right. If we decide to not make this the default behavior then this will be hard to change. -> I'll remove this call and make it user callable.

Reply:
addressed in https://github.com/aws/s2n-tls/pull/3216/commits/8e5302f84c4df8f4aaaa2fb4bdb32c0d753fb965